Engine/src/neat/neat-neural-network.cpp

```diff
@@ -0,0 +1,125 @@
+                    inputs.push_back(NeuronInput(link.GetInId(), link.GetWeight()));
+                }
+            }
+            ffneurons.push_back(FeedForwardNeuron(neuron.GetId(), neuron.GetBias(), inputs));
+        }
+    }
+    ffneurons_ = ffneurons;
+}
+
+std::vector<double> NeuralNetwork::Activate(std::vector<double> input_values) const {
+    assert(input_values.size() == input_ids_.size());
+    std::unordered_map<int, double> values;
+    for (int i = 0; i < input_values.size(); i++) {
+        values[input_ids_[i]] = input_values[i];
+    }
+    /*for (int i = 0; i < output_ids_.size(); i++) {
+        values[output_ids_[i]] = 0;
+    }*/
+    for (const FeedForwardNeuron &ffneuron:ffneurons_) {
+        if (values.find(ffneuron.id) == values.end()) {//if ffneuron is not already activated
+            double value = 0;
+            for (const NeuronInput &input:ffneuron.inputs) {
+                assert(values.find(input.input_id) != values.end()); //previous neurons have to be activated
+                value += values[input.input_id] * input.weight;
+            }
+            value+= ffneuron.bias;
+            value = activation_funtion(value);
+            values[ffneuron.id] = value;
+        }
+    }
+    std::vector<double> result;
+    for (int i  = 0; i < output_ids_.size(); i++) {
+        result.push_back(values[output_ids_[i]]);
+    }
+    return result;
+
+}
+
+std::vector<std::vector<Neuron> > get_layers(Genome &genom) {
+    std::vector<std::vector<Neuron> > layers;
+    std::vector<Neuron> input_layer;
+    std::vector<int> active; // a neuron is active if all of its incoming neurons belong to an active layer. A layer is active if all of its neurons are active.
+                             //input neurons are also active and they form an active layer
+    std::vector<Neuron> neurons = genom.GetNeurons();
+    std::vector<Link> links = genom.GetLinks();
+
+    for (int i = 0; i < genom.GetInputCount(); i++) {
+        int inputid = neurons[i].GetId();
+        input_layer.push_back(neurons[i]);
+        active.push_back(inputid);
+    }
+    layers.push_back(input_layer);
+    std::vector<Neuron> output_layer;
+    for (int i = genom.GetInputCount(); i < genom.GetInputCount() + genom.GetOutputCount(); i++) {
+        output_layer.push_back(neurons[i]);
+    }
+    int N_neurons = neurons.size(); //total number of neurons in the network
+    while (active.size() < N_neurons - genom.GetOutputCount()) {
+        std::vector<Neuron> layer;
+        for (int i = genom.GetInputCount() + genom.GetOutputCount(); i < N_neurons;  i++){ //we dont check for the input neurons, they are already added
+            Neuron neuron = neurons[i];
+            if (!contains(active, neuron.GetId())) {
+                bool is_active = true;
+                for (const Link& link: links) {
+                    if (link.IsActive() && link.GetOutId() == neuron.GetId() && !contains(active, link.GetInId())) {
```

Comment:
At some point this will be time-consuming, the complexity seems to be O(active.size() * links.size() * genom.size() * number of layers). But I think we can keep this for now, we can fix this (and other places) later when we need to optimize